gohcl/decode.go

```diff
@@ -31,27 +60,51 @@ import (
 // may still be accessed by a careful caller for static analysis and editor
 // integration use-cases.
 func DecodeBody(body hcl.Body, ctx *hcl.EvalContext, val interface{}) hcl.Diagnostics {
+	return global.DecodeBody(body, ctx, val)
+}
+
+// DecodeBody extracts the configuration within the given body into the given
+// value. This value must be a non-nil pointer to either a struct or
+// a map, where in the former case the configuration will be decoded using
+// struct tags and in the latter case only attributes are allowed and their
+// values are decoded into the map.
+//
+// The given EvalContext is used to resolve any variables or functions in
+// expressions encountered while decoding. This may be nil to require only
+// constant values, for simple applications that do not support variables or
+// functions.
+//
+// The returned diagnostics should be inspected with its HasErrors method to
+// determine if the populated value is valid and complete. If error diagnostics
+// are returned then the given value may have been partially-populated but
+// may still be accessed by a careful caller for static analysis and editor
+// integration use-cases.
+func (d *Decoder) DecodeBody(body hcl.Body, ctx *hcl.EvalContext, val interface{}) hcl.Diagnostics {
 	rv := reflect.ValueOf(val)
 	if rv.Kind() != reflect.Ptr {
 		panic(fmt.Sprintf("target value must be a pointer, not %s", rv.Type().String()))
 	}
 
-	return decodeBodyToValue(body, ctx, rv.Elem())
+	return d.decodeBodyToValue(body, ctx, rv.Elem())
 }
 
-func decodeBodyToValue(body hcl.Body, ctx *hcl.EvalContext, val reflect.Value) hcl.Diagnostics {
+func (d *Decoder) decodeBodyToValue(body hcl.Body, ctx *hcl.EvalContext, val reflect.Value) hcl.Diagnostics {
 	et := val.Type()
 	switch et.Kind() {
 	case reflect.Struct:
-		return decodeBodyToStruct(body, ctx, val)
+		return d.decodeBodyToStruct(body, ctx, val)
 	case reflect.Map:
-		return decodeBodyToMap(body, ctx, val)
+		return d.decodeBodyToMap(body, ctx, val)
 	default:
 		panic(fmt.Sprintf("target value must be pointer to struct or map, not %s", et.String()))
 	}
 }
 
-func decodeBodyToStruct(body hcl.Body, ctx *hcl.EvalContext, val reflect.Value) hcl.Diagnostics {
+func (d *Decoder) decodeBodyToStruct(body hcl.Body, ctx *hcl.EvalContext, val reflect.Value) hcl.Diagnostics {
+	if fn, ok := d.bodyConvertors[val.Type()]; ok {
+		return fn(body, ctx, val.Addr().Interface())
+	}
```

Comment:
Nomad allows for constraint blocks where certain `operators` can be expressed as LHS constants. For example, the canonical `distinct_hosts` constraint is written as follows:

```hcl
constraint {
  operator  = "distinct_hosts"
  value     = "true"
}
```
However, Nomad supports an alternative representation of this same constraint.

```hcl
constraint {
    distinct_hosts = true
}
```
Before re-adding this code, the shortened form would produce the following error.

```text
Unsupported argument; An argument named "distinct_hosts" is not expected here.
```

Fixes https://github.com/hashicorp/nomad/pull/18122#issuecomment-1813537980